conan/tools/cmake/toolchain/toolchain.py

```diff
@@ -240,14 +241,21 @@ def generate(self):
         cache_variables = {}
         for name, value in self.cache_variables.items():
             if isinstance(value, bool):
-                cache_variables[name] = "ON" if value else "OFF"
+                cache_variables[name] = {"value": "ON" if value else "OFF", "type": "BOOL"}
+            elif isinstance(value, Path):
+                cache_variables[name] = {
+                    "value": str(value),
+                    "type": "FILEPATH" if value.is_file() else 'PATH'
+                }
+            elif isinstance(value, str):
+                cache_variables[name] = {"value": value, "type": "STRING"}
             elif isinstance(value, _PackageOption):
                 if str(value).lower() in ["true", "false", "none"]:
-                    cache_variables[name] = "ON" if bool(value) else "OFF"
+                    cache_variables[name] = {"value": "ON" if bool(value) else "OFF", "type": "BOOL"}
                 elif str(value).isdigit():
                     cache_variables[name] = int(value)
                 else:
-                    cache_variables[name] = str(value)
+                    cache_variables[name] = {"value": str(value), "type": "STRING"}
```

Comment:
Replace cache_variables with an object with both `value` and `type` where appropriate

Test example: 

```
tc = CMakeToolchain(self)
tc.cache_variables["MY_BOOL"] = False
tc.cache_variables["MY_FILE_PATH"] = Path("/path/to/file")   # assert Path("/path/to/file").is_file()
tc.cache_variables["MY_FILE_PATH"] = Path("/path/to/dir")   # assert Path("/path/to/file").is_dir()
tc.generate()
```